### DISCUSSION TITLE
Remove : char from certain field names in atc.audit logs

### DISCUSSION THREAD

**Discussion by vigneshvpra** [concourse/concourse#6193] [Help & Support]
## Steps to reproduce

1. Set CONCOURSE_ENABLE_SYSTEM_AUDITING=true 
2. Use fly set-pipeline to make changes to a pipeline
3. A log event is generated like - Oct 21 19:53:46 concourse-web-stage-abc concourse-web[1111]: {"timestamp":"2020-10-21T19:53:46.953938877Z","level":"info","source":"atc","message":"atc.audit","data":{"action":"SaveConfig","parameters":{":pipeline_name":["test-pipeline"],":team_name":["test"]},"user":"testuser"}}
4. The ":pipeline_name" and ":team_name" fields have a : in front of them

## Expected results
These 2 fields should just be "pipeline_name" and "team_name"

## Actual results
These 2 fields have a : in front of them

## Triaging info

* Concourse version: 6.5.1
* Browser (if applicable):
* Did this used to work?
6.3.0 also had the same issue. Haven't checked before that.

**Comment by jamieklassen** [comment]
I'm having trouble understanding why you consider this a bug -- was there a version of concourse prior to 6.3.0 where these fields _didn't_ have the colons in the logs? do you have some system consuming these logs that can't handle the colons?

**Reply by vigneshvpra** [reply]
@jamieklassen I am not sure whether this issue was present prior to 6.3.0. Our logging system is unable to handle colons in the field names and they end up getting dropped altogether. Is there a reason why only those 2 fields are prefixed with a :? Thanks

**Reply by jamieklassen** [reply]
I did some research for you.
It's an artifact of the libraries we depend on. The auditor itself logs based on the form parameters of the request:

https://github.com/concourse/concourse/blob/5edc00f848aa6c3659b25c1900e1be31a6c82b4e/atc/auditor/auditor.go#L168

and since our API is constructed using https://github.com/tedsuo/rata:

https://github.com/concourse/concourse/blob/5edc00f848aa6c3659b25c1900e1be31a6c82b4e/atc/api/handler.go#L223

and that library itself depends on http://github.com/bmizerany/pat:

https://github.com/tedsuo/rata/blob/07d200713958d05692ac35aa8b5f1949342fee80/router.go#L32

and `pat` automatically populates the URL parameters for requests with variables from the URL path pattern:

https://github.com/bmizerany/pat/blob/6226ea591a40176dd3ff9cd8eff81ed6ca721a00/mux.go#L115

the issue really comes from the fact that `pat`'s URL captures are denoted by prefixing with a colon.

We've had this format since the feature was introduced so I'm mildly wary of forcing a breaking change (since audit tends to be an important and slow-moving part of most organizations), but it also feels like a pretty awkward thing to have a flag for.

However this feature was implemented in a pretty minimal way and what you're suggesting does seem like a _better_ default. What is your logging system? Will other people hit this problem?

**Reply by jamieklassen** [reply]
FWIW I'm a bit interested in this now because my PR #6134 is centered around switching away from this library.

**Reply by vigneshvpra** [reply]
We use Elasticsearch and Kibana. It is unusual for field names to start with : so if they are checking if field names must not contain certain characters, they might run into this issue. Would the fix for #6134 fix this issue as well? Thanks for taking a look.

**Reply by jamieklassen** [reply]
Actually `gorilla/mux` uses this `requestWithVars` helper to add data about matched url captures

https://github.com/gorilla/mux/blob/98cb6bf42e086f6af920b965c38cacc07402d51b/mux.go#L198

and it relies on populating a value in the request context

https://github.com/gorilla/mux/blob/98cb6bf42e086f6af920b965c38cacc07402d51b/mux.go#L447

and doesn't use form parameters at all. This means that with my change, the `:pipeline_name` and `:team_name` fields in the logs would disappear entirely, and my implementation would effectively break the auditing feature! 😬 thanks for bringing my attention to it!! Anyway this feels like a pretty reasonable feature request to me.

**Reply by vigneshvpra** [reply]
Thanks @jamieklassen! Let me know if you need anything else.
